Exposure support

scholtzan opened the pull request:
Fixes https://github.com/mozilla/jetstream/issues/654 and https://github.com/mozilla/jetstream/issues/652 and https://github.com/mozilla/jetstream/issues/643

What this currently does:

Depending on whether enrollment or exposure-based metrics are available, jetstream will create separate metric tables which are named like `<slug>_<enrollment|exposures>_<period>_<n>`. The data from these metric tables will be combined and used for computing statistics which will get stored in single statistics tables per analysis period, which are then used to create the views that are exported to GCS. Metadata that is exported to GCS also contains information about the `analysis_basis`. So Experimenter can use that information to display enrollments/exposure-based results.

One reason for this approach was to maintain compatibility with the current Experimenter implementation. But I'm open to make (breaking?) changes here. Alternatives that might also work here:
* have separate tables and exported views for enrollments and exposures
* have a `analysis_basis` segment (not of boolean type but a string for `exposure`/`enrollments`)
* prefix the column names with exposure/enrollment in the tables

scholtzan commented on jetstream/metric.py:
```diff
@@ -0,0 +1,53 @@
+
+
+@attr.s(auto_attribs=True, frozen=True, slots=True)
+class Metric:
```
`class Metric(mozanalysis.metrics.Metric)` doesn't work. mypy complains that the fields defined in the mozanalysis `Metrics` class are missing and when passing `Metric` to a Dask worker the `analysis_basis` value gets erased.

tdsmith replied:
That's fine; I think we don't really want to deal with cross-package inheritance here anyway.

scholtzan commented on jetstream/config.py:
```diff
@@ -363,28 +422,36 @@ class MetricDefinition:
+            metric = Metric.from_mozanalysis_metric(
+                mozanalysis_metric=mozanalysis_metric,
+                analysis_basis=self.analysis_basis
+                or mozanalysis.experiment.AnalysisBasis.ENROLLMENTS,
```
The default is to use enrollments for analysis by default. Based on the proposal it sounds like exposures should become the default at some point. I guess that would break current assumptions of the results and likely requires some communication to parties that use the results.

scholtzan commented on jetstream/analysis.py:
```diff
@@ -216,22 +238,30 @@ def calculate_statistics(
+        self, segment_data: DataFrame, segment: str, analysis_basis: AnalysisBasis
+    ) -> StatisticResultCollection:
         """Count and missing count statistics."""
+        metric = analysis_basis.value + "_identity"
```
For `counts`, this will take the analysis basis into account. So if there are metrics that are based on enrollments and some based on exposures, then the statistics result table will have `enrollments_identity` and `exposures_identity` metrics for each branch

tdsmith replied:
this is a breaking change for the console so we need to figure out how to handle that before merging (maybe revisiting the adr)

scholtzan replied:
actually, renaming is not necessary anymore now that we have an analysis_basis column

scholtzan commented:
I made the changes based on the [ADR](https://github.com/mozilla/jetstream/blob/main/docs/adr-0002-publishing_statistics_with_analysis_basis.md) and added an `analysis_basis` column to the statistics result tables

tdsmith commented on jetstream/config.py:
```diff
@@ -364,28 +366,32 @@ class MetricDefinition:
                 description=self.description,
                 bigger_is_better=self.bigger_is_better,
+                analysis_basis=self.analysis_basis
+                or mozanalysis.experiment.AnalysisBasis.ENROLLMENT,
```
In general I think it'll be most useful to assume that most metrics should be computed on an exposure basis if an experiment has a useful concept of exposure. Not all of them will.

I think it also sounds confusing if the default basis is different depending on whether or not we've defined something to be true about the experiment.

It sounds like maybe we should back up and figure out what we'd like the configuration files to look like in each case.

We also need to make sure we can communicate the basis of our metrics to the console. Do we want our days_of_use metric to be sometimes computed on an enrollment basis and sometimes computed on an exposure basis, depending on the experiment? Like, maybe, but I dunno. (This _might_ be a reason to revisit the ADR: if we require metrics computed for different bases to have distinct names, that makes it harder to connect the exposure-DOU to the enrollment-DOU, and we will care about both numbers. It might need to be possible to specify bases as ["exposure", "enrollment"], and to introduce a breaking change to the contract of the stats tables so that you need to group by the basis as well. That would probably require a week of data science feedback time but I don't think that's intolerable.)

scholtzan replied:
I added some updates to the ADR in https://github.com/mozilla/jetstream/pull/724/

scholtzan replied:
For now I implemented option 1, which seems more convenient when working with the statistics results even if it might break the contract. I'm not sure if there are data scientists that actually work with the statistics tables directly that would be impacted or if it would only affect Experimenter console?

tdsmith replied:
Su has some redash dashboards; I'll set up a PR to DTMO and circulate it as the notification.

tdsmith commented on jetstream/analysis.py:
```diff
@@ -166,7 +175,7 @@ def calculate_metrics(
         Returns the BigQuery table results are written to.
         """
-
+        print(self.config)
```
I see this was just a wip commit, but: 🖨️🖨️🖨️ :D

tdsmith commented on jetstream/analysis.py:
```diff
@@ -120,17 +122,32 @@ def _get_timelimits_if_ready(
+    def _table_name(
+        self, window_period: str, window_index: int, analysis_basis: Optional[AnalysisBasis] = None
+    ) -> str:
         assert self.config.experiment.normandy_slug is not None
```
might be worth a docstring here explaining that aggregates tables should have an AnalysisBasis but statistics tables shouldn't

tdsmith commented on jetstream/config.py:
```diff
@@ -316,6 +343,28 @@ def _validate_yyyy_mm_dd(instance: Any, attribute: Any, value: Any) -> None:
+
+    name: str
+    data_source: DataSourceReference
+    select_expr: str
```
select_expression to be consistent with Metrics

tdsmith commented on jetstream/metric.py:
```diff
@@ -6,22 +6,35 @@
 
+    name: str
+    data_source: mozanalysis.metrics.DataSource
+    select_expr: str
```
probably good to be consistent internally and spell out "expression"

tdsmith reviewed (commented):
let's chat monday; i think this exposes a few questions about how we want the config to look and how we want to handle experiments without an exposure basis vs experiments with an exposure basis that we should probably try to address head-on

scholtzan commented on jetstream/config.py:
```diff
@@ -364,28 +366,32 @@ class MetricDefinition:
     friendly_name: Optional[str] = None
     description: Optional[str] = None
     bigger_is_better: bool = True
+    analysis_basis: Optional[mozanalysis.experiment.AnalysisBasis] = None
```
Added support for having a single analysis basis or a list of multiple bases

scholtzan commented:
This now reflects what is described in the ADR. If we keep enrollments as the default analysis basis then this will not introduce any breaking changes for Experimenter (analysis_basis would simply get ignored as it's now). As soon as we have experiments that use exposures we should probably indicate those in the Experimenter console frontend.

scholtzan commented on jetstream/__init__.py:
```diff
@@ -1,6 +1,8 @@
 import enum
 import re
 
+STATISTICS_SCHEMA_VERSION = 2
```
Depends on https://github.com/mozilla/jetstream/pull/726

tdsmith commented on jetstream/analysis.py:
```diff
@@ -439,22 +476,64 @@ def run(self, current_date: datetime, dry_run: bool = False) -> None:
+                "num_exposure_events",
+            ] + segment_labels
+
+            metrics_data = reduce(
```
I think this makes a single table that has both exposure- and enrollment-basis data in it. We need to calculate the statistics over the exposure-basis data and the enrollment-basis data separately. Keeping the tables separate might make more sense.

Anyway, if a metric has more than one basis won't the column names conflict?

scholtzan replied:
Indeed, the computations need to be done separately. Actually that simplifies things...

tdsmith reviewed (commented):
Lots going on in here. In the future it might be worth thinking about how to sequence changes like this as a set of smaller PRs.

The configuration aspect of this generally makes sense to me for now, but we will need to figure out the migration path towards making exposures the default, pretty soon.

I'm confused about the analysis part; let me know if my comments are confusing and let's jump on a call.

scholtzan commented on jetstream/metric.py:
```diff
@@ -0,0 +1,54 @@
+    friendly_name: Optional[str] = None
+    description: Optional[str] = None
+    bigger_is_better: bool = True
+    analysis_basis: Union[
```
Yes, that would make a lot of things so much easier!

scholtzan commented on jetstream/config.py:
```diff
@@ -325,6 +386,8 @@ class ExperimentSpec:
     segments: List[SegmentReference] = attr.Factory(list)
     skip: bool = False
+    exposure_signal: Optional[ExposureSignalDefinition] = None
+    exposure_query: Optional[str] = None
```
mozanalysis would indeed throw an error, but I think going with option 1 is probably good enough. If some need for having custom queries arises we can do option 2

scholtzan commented on jetstream/analysis.py:
```diff
@@ -439,34 +473,57 @@ def run(self, current_date: datetime, dry_run: bool = False) -> None:
+                    )
+                    continue
+
+                segment_labels = ["all"] + [s.name for s in self.config.experiment.segments]
+                for segment in segment_labels:
+                    segment_data = self.subset_to_segment(segment, metrics_dataframe)
+                    for m in self.config.metrics[period]:
+                        segment_results += self.calculate_statistics(
+                            m,
+                            segment_data,
+                            segment,
+                            analysis_basis,
+                        ).to_dict()["data"]
+
+                        segment_results += self.counts(
+                            segment_data, segment, analysis_basis
+                        ).to_dict()["data"]
+
+        results.append(
+            self.save_statistics(
+                period,
+                segment_results,
+                self._table_name(period.value, len(time_limits.analysis_windows)),
+            )
+        )
```
I guess we can just compute segments for each analysis basis separately and then store the results to the statistics table all together.

scholtzan commented:
I updated this PR to use `analysis_bases` instead of `analysis_basis`, which simplified the code quite a bit. I also updated the statistics computation logic.

tdsmith commented on jetstream/metadata.py:
```diff
@@ -15,9 +15,10 @@
 class MetricsMetadata:
-    friendly_name: str
-    description: str
+    friendly_name: Optional[str]
+    description: Optional[str]
```
This seems realistic; I'm not sure how this typed before :D but are we sure that Experimenter can accept missingness here? @emtwo

scholtzan replied:
Looking at the Experimenter code [here](https://github.com/mozilla/experimenter/blob/258574f9f90abae0af292f51efc5cbe8f9fcd102/app/experimenter/nimbus-ui/src/lib/visualization/types.ts#L23) and [here](https://github.com/mozilla/experimenter/blob/c121f12957ff79e9233d4de4ac83851b1a0fcbf9/app/experimenter/nimbus-ui/src/components/AppLayoutSidebarLaunched/index.tsx#L48) it appears both are expected to be there

scholtzan replied:
I changed the types here to `str` and made sure if values are missing they are replaced by empty strings so that Experimenter doesn't break

tdsmith replied:
🌈 perfect thanks

tdsmith reviewed (approved):
This looks great, and indeed much cleaner; thanks for all your patience here.

I think what will happen after we merge this is:
- All metrics will be assumed to be enrollment-basis by default
- We'll start producing enrollment-flavoured aggregates tables with different names
- We'll add a new column to the statistics tables but functionally nothing will change

Once we add our first exposure-basis metric:
- We'll additionally start producing exposure-flavoured aggregates tables
- We'll actually break the old contract for the statistics tables

I think this means we should not actually allow an exposure-basis metric to be instantiated until Experimenter can handle it without barfing. Probably the next step is that Experimenter should just start filtering for enrollment-basis stats and ignore the rest for now.

Can we make it an error to instantiate an exposure basis metric in a way that allows the tests to still exercise useful functionality? Some kind of feature-flag that we monkeypatch in the tests? Then we can file a bug to flip the flag that's blocked on Experimenter support for the exposure metrics.

scholtzan commented:
I changed the code to raise an exception when exposure analysis basis is used. I monkeypatched the tests so they still run successfully.

scholtzan commented:
I filed https://github.com/mozilla/jetstream/issues/796 and https://github.com/mozilla/experimenter/issues/5827 for flipping exposure support
